# Homework-3 

## Use the Titanic Dataset to Practice Data Wrangling Tasks
### Part 1: Import libaries , load the data, inspect the data, then get the name of the variables and create a small markdown table  with variable names and types (e.g numerical or categorical ordinal, categorical nominal etc.)

### Part 2: 
    - Impute missing values for "Embarked" with the mode (most frequent value)
    - Impute missing value for Age using two following strategies
        - A) Simple mean imputation (fill with the overal mean age)
        - B) Group-wise median imputation: for each combination of Sex and Pclass compute the median age. Impute missing age using the median age of passenger's sex and Pclass group

### Part 3: Create a FamilySize variable (FamilySize = SibSp + Parch +1) and compute:
    - Minimum, Maximum, and mean FamilySize
    - How many passengers were traveling alone
    
### Part 4:  Create a filtered dataframe of all passengers:
    - Who are female
    - Who are male 
    - Who are under 16 years old   
    - Compute the survival rate for the above filtered data
    - Compute survival rate by Sex and Pclass (use groupby and mean) and present the result in a markedown table (rows: sex, columns: Pclass) or using print formating.
    - Interpret your result in 2-3 sentences
     
    
### Part 5: Compute the average fare for each Pclass (Fare values are per tickets not per person (there coud be more than 1 person on a ticket) so make sure to account for that. 

## Part 1)
Import libaries , load the data, inspect the data, then get the name of the variables and create a small markdown table with variable names and types (e.g numerical or categorical ordinal, categorical nominal etc.)

In [107]:
#import libraries
import numpy as py
import pandas as pd

In [130]:
#load data; ti is short for titanic
ti = pd.read_csv('titanic.csv')

In [109]:
ti.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [131]:
ti.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Variable Types


| Variable | Type |
|:-------------:|:---------------:|
| PassengerID |Cat, nominal|
| Survived | Cat, Nominal |
| Pclass | Cat, Ordinal |
| Name | Cat, Nominal |
| Sex | Cat, Nominal |
| Age | Num, Cont |
| SibSp | Num, Discrete |
| Parch | Num, Discrete |
| Ticket | Cat, Nominal |
| Fare | Cat, Cont |
| Cabin | Cat, Nominal |
| Embarked | Cat, Nominal |

## Part 2)

### Imputting Embarked Missing Data

In [111]:
# finding the most common value 
emb = ti['Embarked'].mode()[0] #most embarked location

In [216]:
# from the previous info summary we know there are two NaN values for embarked
ti2 = ti.copy() # making a new dataframe in case I want to see the original for some reason
ti2['Embarked'] = ti2['Embarked'].fillna('emb')
ti2.info() # check to see if the non-null count matches the number of entries, yay it does

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Age: Mean Method

In [133]:
allmean = ti2.copy()
allmean['Age'] = allmean['Age'].fillna(allmean['Age'].mean())

In [134]:
allmean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Age: Median for Sex and Pclass

In [140]:
# I realized these medians needed to go in the main dataframe and tried to do that without making six separate frames per gender-class,
# but I could not figure how, so I did each gender-class separately and then stacked the results back together.
first_female = ti2[((ti2['Sex'] == 'female') & (ti2['Pclass'] == 1))].copy()
first_female.head()

PassengerId  Survived  Pclass  \
1             2         1       1   
3             4         1       1   
11           12         1       1   
31           32         1       1   
52           53         1       1   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
11                            Bonnell, Miss Elizabeth  female  58.0      0   
31     Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
52           Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   

    Parch    Ticket      Fare Cabin Embarked  
1       0  PC 17599   71.2833   C85        C  
3       0    113803   53.1000  C123        S  
11      0    113783   26.5500  C103        S  
31      0  PC 17569  146.5208   B78        C  
52      0  PC 17572   76.7292   D33        C

In [225]:
#first_female.info() #see if there are NaN values

In [202]:
first_female['Age'] = first_female['Age'].fillna(first_female['Age'].median())
first_female.head()
first_female.info() # non-null count to see check if it worked

<class 'pandas.core.frame.DataFrame'>
Index: 94 entries, 1 to 887
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  94 non-null     int64  
 1   Survived     94 non-null     int64  
 2   Pclass       94 non-null     int64  
 3   Name         94 non-null     object 
 4   Sex          94 non-null     object 
 5   Age          94 non-null     float64
 6   SibSp        94 non-null     int64  
 7   Parch        94 non-null     int64  
 8   Ticket       94 non-null     object 
 9   Fare         94 non-null     float64
 10  Cabin        81 non-null     object 
 11  Embarked     94 non-null     object 
dtypes: float64(2), int64(5), object(5)
memory usage: 9.5+ KB


In [180]:
sec_female = ti2[((ti2['Sex'] == 'female') & (ti2['Pclass'] == 2))].copy()
sec_female.head()
sec_female.info() 

<class 'pandas.core.frame.DataFrame'>
Index: 76 entries, 9 to 880
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  76 non-null     int64  
 1   Survived     76 non-null     int64  
 2   Pclass       76 non-null     int64  
 3   Name         76 non-null     object 
 4   Sex          76 non-null     object 
 5   Age          74 non-null     float64
 6   SibSp        76 non-null     int64  
 7   Parch        76 non-null     int64  
 8   Ticket       76 non-null     object 
 9   Fare         76 non-null     float64
 10  Cabin        10 non-null     object 
 11  Embarked     76 non-null     object 
dtypes: float64(2), int64(5), object(5)
memory usage: 7.7+ KB


In [204]:
sec_female['Age'] = sec_female['Age'].fillna(sec_female['Age'].median())
#sec_female.info()

<class 'pandas.core.frame.DataFrame'>
Index: 76 entries, 9 to 880
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  76 non-null     int64  
 1   Survived     76 non-null     int64  
 2   Pclass       76 non-null     int64  
 3   Name         76 non-null     object 
 4   Sex          76 non-null     object 
 5   Age          76 non-null     float64
 6   SibSp        76 non-null     int64  
 7   Parch        76 non-null     int64  
 8   Ticket       76 non-null     object 
 9   Fare         76 non-null     float64
 10  Cabin        10 non-null     object 
 11  Embarked     76 non-null     object 
dtypes: float64(2), int64(5), object(5)
memory usage: 7.7+ KB


In [205]:
third_female = ti2[((ti2['Sex'] == 'female') & (ti2['Pclass'] == 3))].copy()
third_female.head()
#third_female.info()

<class 'pandas.core.frame.DataFrame'>
Index: 144 entries, 2 to 888
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  144 non-null    int64  
 1   Survived     144 non-null    int64  
 2   Pclass       144 non-null    int64  
 3   Name         144 non-null    object 
 4   Sex          144 non-null    object 
 5   Age          102 non-null    float64
 6   SibSp        144 non-null    int64  
 7   Parch        144 non-null    int64  
 8   Ticket       144 non-null    object 
 9   Fare         144 non-null    float64
 10  Cabin        6 non-null      object 
 11  Embarked     144 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 14.6+ KB


In [207]:
third_female['Age'] = third_female['Age'].fillna(third_female['Age'].median())
#third_female.info()

In [209]:
first_male = ti2[((ti2['Sex'] == 'male') & (ti2['Pclass'] == 1))].copy()
first_male.info()

<class 'pandas.core.frame.DataFrame'>
Index: 122 entries, 6 to 889
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  122 non-null    int64  
 1   Survived     122 non-null    int64  
 2   Pclass       122 non-null    int64  
 3   Name         122 non-null    object 
 4   Sex          122 non-null    object 
 5   Age          101 non-null    float64
 6   SibSp        122 non-null    int64  
 7   Parch        122 non-null    int64  
 8   Ticket       122 non-null    object 
 9   Fare         122 non-null    float64
 10  Cabin        95 non-null     object 
 11  Embarked     122 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 12.4+ KB


In [229]:
first_male['Age'] = first_male['Age'].fillna(first_male['Age'].median())
#first_male.info()

In [218]:
sec_male = ti2[((ti2['Sex'] == 'male') & (ti2['Pclass'] == 2))].copy()
sec_male.head()
sec_male.info() 

<class 'pandas.core.frame.DataFrame'>
Index: 108 entries, 17 to 886
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  108 non-null    int64  
 1   Survived     108 non-null    int64  
 2   Pclass       108 non-null    int64  
 3   Name         108 non-null    object 
 4   Sex          108 non-null    object 
 5   Age          99 non-null     float64
 6   SibSp        108 non-null    int64  
 7   Parch        108 non-null    int64  
 8   Ticket       108 non-null    object 
 9   Fare         108 non-null    float64
 10  Cabin        6 non-null      object 
 11  Embarked     108 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 11.0+ KB


In [230]:
sec_male['Age'] = sec_male['Age'].fillna(sec_male['Age'].median())
#sec_male.info() 

In [227]:
third_male = ti2[((ti2['Sex'] == 'male') & (ti2['Pclass'] == 3))].copy()
#third_male.head()
third_male.info() 

In [226]:
third_male['Age'] = third_male['Age'].fillna(third_male['Age'].median())
#third_male.info()

In [ ]:
# stack the gender and class dataframes back ontop of each other
median_age = pd.concat([first_female, sec_female, third_female, first_male, sec_male, third_male])
median_age

## Part 3

In [231]:
ti3 = ti2.copy()
ti3['FamilySize'] = ti3['SibSp'] + ti3['Parch'] + 1 
ti3.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  FamilySize  
0      0         A/5 21171   7.2500   NaN        S           2  
1      0          PC 17599  71.2833   C85        C           2  
2      0  STON/O2. 3101282   7.9250   NaN        S           1  
3      0            113803  53.1000  C123        S           2  
4      0            373450   8.0500   NaN        S           1

In [241]:
ti3['FamilySize'].describe()

count    891.000000
mean       1.904602
std        1.613459
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max       11.000000
Name: FamilySize, dtype: float64

In [ ]:
#mean is 1.90, min is 1, max is 11

In [255]:
alone = ti3.groupby('FamilySize')['FamilySize'].agg('count')
alone

FamilySize
1     537
2     161
3     102
4      29
5      15
6      22
7      12
8       6
11      7
Name: FamilySize, dtype: int64

In [254]:
print(f'{alone[1]} people are traveling alone')

537 people are traveling alone


## Part 4

In [263]:
female = ti3[(ti3['Sex'] == 'female')].copy()
female.head()
male = ti3[(ti3['Sex'] == 'male')].copy()
male.head()
child = ti3[(ti3['Age'] < 16)].copy() # I an interpretting the assignment's question 'under sixteen' to not include 16 yearolds
child.head()

PassengerId  Survived  Pclass                                 Name  \
7             8         0       3        Palsson, Master Gosta Leonard   
9            10         1       2  Nasser, Mrs. Nicholas (Adele Achem)   
10           11         1       3       Sandstrom, Miss Marguerite Rut   
14           15         0       3  Vestrom, Miss Hulda Amanda Adolfina   
16           17         0       3                  Rice, Master Eugene   

       Sex   Age  SibSp  Parch   Ticket     Fare Cabin Embarked  FamilySize  
7     male   2.0      3      1   349909  21.0750   NaN        S           5  
9   female  14.0      1      0   237736  30.0708   NaN        C           2  
10  female   4.0      1      1  PP 9549  16.7000    G6        S           3  
14  female  14.0      0      0   350406   7.8542   NaN        S           1  
16    male   2.0      4      1   382652  29.1250   NaN        Q           6

In [283]:
status_f = female.groupby('Survived')['Survived'].agg('count')
alive_f = round(float(status_f[1])/len(female['Name']),4)
print(f'proportion of females who survived is {alive_f}')

proportion of females who survived is 0.742


In [284]:
status_m = male.groupby('Survived')['Survived'].agg('count')
alive_m = round(float(status_m[1])/len(male['Name']),4)
print(f'proportion of males who survived is {alive_m}')

proportion of males who survived is 0.1889


In [285]:
status_ch = child.groupby('Survived')['Survived'].agg('count')
alive_ch = round(float(status_ch[1])/len(child['Name']),4)
print(f'proportion of children under 16 who survived is {alive_ch}')

proportion of children under 16 who survived is 0.5904


In [286]:
#  Compute survival rate by Sex and Pclass (use groupby and mean) 
# I do not understand what mean has to do with this. 
# I've already filtered for Sex and Pclass in part 2 and can easily calculate survival rate from those dataframes


In [290]:
f1_life = first_female.groupby('Survived')['Survived'].agg('count')
f1_alive = round(float(f1_life[1])/len(first_female['Name']),4)
f1_alive

f2_life = sec_female.groupby('Survived')['Survived'].agg('count')
f2_alive = round(float(f2_life[1])/len(sec_female['Name']),4)
f2_alive

f3_life = third_female.groupby('Survived')['Survived'].agg('count')
f3_alive = round(float(f3_life[1])/len(third_female['Name']),4)
f3_alive

0.5

In [293]:
m1_life = first_male.groupby('Survived')['Survived'].agg('count')
m1_alive = round(float(m1_life[1])/len(first_male['Name']),4)
m1_alive

m2_life = sec_male.groupby('Survived')['Survived'].agg('count')
m2_alive = round(float(m2_life[1])/len(sec_male['Name']),4)
m2_alive

m3_life = third_male.groupby('Survived')['Survived'].agg('count')
m3_alive = round(float(m3_life[1])/len(third_male['Name']),4)
m3_alive

0.1354

### Survival Ratios
| Sex | 1st Class | 2nd Class | 3rd Class |
|:---:|:---------:|:---------:|:---------:|
| F | 96.81% | 92.11% | 50.00%| 
| M | 36.89% | 15.74% | 13.54%| 

Women had a far higher survival rate than men, even as survival rate decreased as passenger class luxary decreased in in both genders. It is interesting to note that women had a more dramatic gap between 2nd and 3rd class, while men had a larger survival gap between 1st and 2nd. 

## Part 5

In [306]:
ti5 = ti3.copy()
ti5

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                                Heikkinen, Miss Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                        Graham, Miss Margaret Edith  female  19.0      0   
888            Johnston, Miss Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  FamilySize  
0        0         A/5 21171   7.2500   NaN        S           2  
1        0          PC 17599  71.2833   C85        C           2  
2        0  STON/O2. 3101282   7.9250   NaN        S           1  
3        0            113803  53.1000  C123        S           2  
4        0            373450   8.0500   NaN        S           1  
..     ...               ...      ...   ...      ...         ...  
886      0            211536  13.0000   NaN        S           1  
887      0            112053  30.0000   B42        S           1  
888      2        W./C. 6607  23.4500   NaN        S           4  
889      0            111369  30.0000  C148        C           1  
890      0            370376   7.7500   NaN        Q           1  

[891 rows x 13 columns]

In [307]:
ticket_count = ti5.groupby('Ticket')['Ticket'].agg('count')
ticket_count

Ticket
110152         3
110413         3
110465         2
110564         1
110813         1
              ..
W./C. 6608     4
W./C. 6609     1
W.E.P. 5734    1
W/C 14208      1
WE/P 5735      2
Name: Ticket, Length: 681, dtype: int64

In [311]:
# there are only 681 tickets, if I can remove the duplicate ticket fares, I can find the mean fare
unique_fare = ti5.drop_duplicates(subset = ['Ticket'], keep = 'first')
unique_fare.sort_values('Ticket')
    

PassengerId  Survived  Pclass  \
257          258         1       1   
262          263         0       1   
110          111         0       1   
430          431         1       1   
366          367         1       1   
..           ...       ...     ...   
86            87         0       3   
235          236         0       3   
92            93         0       1   
219          220         0       2   
540          541         1       1   

                                                 Name     Sex   Age  SibSp  \
257                               Cherry, Miss Gladys  female  30.0      0   
262                                 Taussig, Mr. Emil    male  52.0      1   
110                    Porter, Mr. Walter Chamberlain    male  47.0      0   
430         Bjornstrom-Steffansson, Mr. Mauritz Hakan    male  28.0      0   
366  Warren, Mrs. Frank Manley (Anna Sophia Atkinson)  female  60.0      1   
..                                                ...     ...   ...    ...   
86                             Ford, Mr. William Neal    male  16.0      1   
235                       Harknett, Miss Alice Phoebe  female   NaN      0   
92                        Chaffee, Mr. Herbert Fuller    male  46.0      1   
219                                Harris, Mr. Walter    male  30.0      0   
540                            Crosby, Miss Harriet R  female  36.0      0   

     Parch       Ticket    Fare Cabin Embarked  FamilySize  
257      0       110152  86.500   B77        S           1  
262      1       110413  79.650   E67        S           3  
110      0       110465  52.000  C110        S           1  
430      0       110564  26.550   C52        S           1  
366      0       110813  75.250   D37        C           2  
..     ...          ...     ...   ...      ...         ...  
86       3   W./C. 6608  34.375   NaN        S           5  
235      0   W./C. 6609   7.550   NaN        S           1  
92       0  W.E.P. 5734  61.175   E31        S           2  
219      0    W/C 14208  10.500   NaN        S           1  
540      2    WE/P 5735  71.000   B22        S           3  

[681 rows x 13 columns]

In [324]:
mean_fare = unique_fare['Fare'].mean()
print(f'mean fare is {round(float(mean_fare),2)}')

mean fare is 23.27
